File: 10/two.py
# --- Advent of code 2021: Day 10 ---

# (Skeleton file automatically created by aocTool, developed by Bolu, 2020-2021.)

lines = open("input.txt").read().splitlines()

open_chars = { "(", "[", "{", "<" }
close_chars = { ")", "]", "}", ">" }
correspondence = { "(": ")", "[": "]", "{": "}", "<": ">" }
points = { ")": 1, "]": 2, "}": 3, ">": 4 }

incomplete_lines = []
scores = []
for line in lines:
	stack = []
	for char in line:
		if char in open_chars:
			stack.append(char)
		else:
			if len(stack) == 0:
				continue
			if char != correspondence[stack.pop()]:
				stack = []
				break
	if len(stack) != 0:
		scores.append(0)
		for char in stack[::-1]:
			scores[-1] = scores[-1] * 5- -points[correspondence[char]]

scores.sort()

print(f"Middle score: {scores[len(scores) // 2]}")
assert scores[len(scores) // 2] == 2116639949

File: 10/two.lua
-- --- Advent of code 2021: Day 10 ---

-- (Skeleton file automatically created by aocTool, developed by Bolu, 2020-2021.)

package.path = package.path .. ";../modules/?.lua"
local tools = require "tools"

lines = tools.readlines("input.txt")

open_chars = {}
open_chars["("] = true
open_chars["{"] = true
open_chars["["] = true
open_chars["<"] = true

correspondence = {}
correspondence["("] = ")"
correspondence["{"] = "}"
correspondence["["] = "]"
correspondence["<"] = ">"

points = {}
points[")"] = 1
points["]"] = 2
points["}"] = 3
points[">"] = 4

incomplete_lines = {}
scores = {}

for _, line in pairs(lines) do
	stack = {}
	for i = 1, #line do
		char = line:sub(i, i)
		if open_chars[char] then
			stack[#stack + 1] = char
		else
			if #stack ~= 0 then
				corresponding = stack[#stack]
				stack[#stack] = nil
				if char ~= correspondence[corresponding] then
					stack = {}
					break
				end
			end
		end
	end
	if #stack ~= 0 then
		scores[#scores + 1] = 0
		for i = 0, #stack - 1  do
			scores[#scores] = scores[#scores] * 5 + points[correspondence[stack[#stack - i]]]
		end
	end
end

table.sort(scores)

print("Middle score: " .. scores[math.ceil(#scores / 2)])
assert(scores[math.ceil(#scores / 2)] == 2116639949)
